### PR TITLE
os/bluestore: verify disk layout of BlueFS

### PR DESCRIPTION
**Pull request by rzarzynski**
This PR targets [bug #25098](https://tracker.ceph.com/issues/25098).

### CODE REVIEW

**Comment by ifed01** [comment]
Just realized one drawback of this patch - it isn't able to detect WAL symlink disappearance. 
May be worth adding?

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -8321,8 +8321,15 @@ void BlueStore::collect_metadata(map<string,string> *pm)
   bdev->collect_metadata("bluestore_bdev_", pm);
   if (bluefs) {
     (*pm)["bluefs"] = "1";
+    // this computed value is for backward compatibility only
+    bool single_shared_device = \
+      !bluefs_layout.dedicated_db && !bluefs_layout.dedicated_wal;
     (*pm)["bluefs_single_shared_device"] = \
```
Comment (rzarzynski) [inline src/os/bluestore/BlueStore.cc]: Do we need backward compatibility here?

**Comment by rzarzynski** [comment]
@ifed01: just added the WAL tracking. Please re-review. Dropping the `needs-qa` label for now.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -8321,8 +8321,15 @@ void BlueStore::collect_metadata(map<string,string> *pm)
   bdev->collect_metadata("bluestore_bdev_", pm);
   if (bluefs) {
     (*pm)["bluefs"] = "1";
+    // this computed value is for backward compatibility only
+    bool single_shared_device = \
+      !bluefs_layout.dedicated_db && !bluefs_layout.dedicated_wal;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: nit: make single_shared_device() helper in bluefs_layout_t?

**Comment by ifed01** [comment]
it's better squash the last commit into the previous one since the parameters it introduces (dedicated_wal/dedicated_db) are referenced from there.

**Comment by rzarzynski** [comment]
> it's better squash the last commit into the previous one  since the parameters it introduces (dedicated_wal/dedicated_db) are  referenced from there.

I think it's just how github sorts the commits (by date; not by the child-parent relationship). In my repo the order looks in following way:
```
$ git log --oneline
11fb60b os/bluestore: verify disk layout of BlueFS.
7c1c2d2 os/bluestore: bluefs_layout_t tracks WAL presence.
7261f83 os/bluestore: introduce bluefs_layout_t.
...
```

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (rzarzynski) [reply src/os/bluestore/BlueStore.cc]: Makes sense.

Comment (rzarzynski) [reply src/os/bluestore/BlueStore.cc]: Added.

**Comment by ifed01** [comment]
LGTM!

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-09-07_14:44:52-rados-wip-kefu-testing-2019-09-07-0159-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-09-08_13:41:17-rados-wip-kefu-testing-2019-09-07-0159-distro-basic-smithi/

**Comment by tchaikov** [comment]
@liewegas shall we backport this change to mimic and nautilus?

**Pull request merged**